src/main/java/fi/aalto/cs/apluscourses/intellij/services/PluginSettings.java

```diff
@@ -176,35 +177,32 @@ public MainViewModel getMainViewModel(@Nullable Project project) {
    */
   public void updateMainViewModel(@Nullable Project project) {
     ProjectKey key = new ProjectKey(project);
-    MainViewModelUpdater updater = mainViewModelUpdaters.get(key);
-    if (updater != null) {
-      updater.restart();
+    var courseProject = courseProjects.get(key);
+    if (courseProject != null) {
+      courseProject.getCourseUpdater().restart();
     }
   }
 
   /**
-   * Creates a main view model and launches an updater for it that runs on a background thread.
-   *
-   * @param project The project to which the created main view model corresponds.
+   * Registers a course project. This creates a main view model. It also starts the updater of the
+   * course project. Calling this method again with the same project has no effect.
    */
-  public void createUpdatingMainViewModel(@NotNull Project project) {
-    ProjectKey key = new ProjectKey(project);
-
-    MainViewModel mainViewModel = mainViewModels.computeIfAbsent(key, projectKey
-        -> {
-      ProjectManager
-          .getInstance()
-          .addProjectManagerListener(project, projectManagerListener);
-      return new MainViewModel(exerciseFilterOptions);
-    });
-
-    mainViewModelUpdaters.computeIfAbsent(key, projectKey -> {
-      MainViewModelUpdater mainViewModelUpdater = new MainViewModelUpdater(
-          mainViewModel, project, MAIN_VIEW_MODEL_UPDATE_INTERVAL,
-          new DefaultNotifier(), IntelliJPasswordStorage::new);
-      mainViewModelUpdater.start();
-      return mainViewModelUpdater;
+  public void registerCourseProject(@NotNull Project project,
```

Comment:
perhaps this could be divided into smaller chunks (like two) and I would love to see a unit (or any other type) of test for this particular method ❤️

Reply:
Good idea!